tests/contrib/django/test_django_snapshots.py

```diff
@@ -1,226 +1,277 @@
     """Runs a django app hosted with a daphne webserver in a subprocess and
     returns a client which can be used to query it.
 
     Traces are flushed by invoking a tracer.shutdown() using a /shutdown-tracer route
     at the end of the testcase.
     """
 
-    # Make sure to copy the environment as we need the PYTHONPATH and _DD_TRACE_WRITER_ADDITIONAL_HEADERS (for the test
-    # token) propagated to the new process.
+    # Make sure to copy the environment as we need the PYTHONPATH.
     env = os.environ.copy()
-    env.update(additional_env or {})
-    assert "_DD_TRACE_WRITER_ADDITIONAL_HEADERS" in env, "Client fixture needs test token in headers"
+    headers = parse_tags_str(env.get("_DD_TRACE_WRITER_ADDITIONAL_HEADERS", ""))
+    headers.update({"X-Datadog-Test-Session-Token": snapshot_token})
+    headers = ",".join(["%s:%s" % (k, v) for k, v in headers.items()])
     env.update(
         {
-            "DJANGO_SETTINGS_MODULE": "tests.contrib.django.django_app.settings",
+            "DJANGO_SETTINGS_MODULE": django_settings,
+            "DD_TRACE_SQLITE3_ENABLED": "0",
+            # Middleware generates large snapshots with redundant info'
+            # For readability we disable it but at least one test case should
+            # override this value to ensure the middleware spans.
+            "DD_DJANGO_INSTRUMENT_MIDDLEWARE": "0",
+            # Database spans can be noisy as operations can happen unrelated
+            # to the endpoint tested.
+            # Tests which access the database should override this flag.
+            "DD_DJANGO_INSTRUMENT_DATABASES": "0",
+            "_DD_TRACE_WRITER_ADDITIONAL_HEADERS": headers,
         }
     )
-
-    # ddtrace-run uses execl which replaces the process but the webserver process itself might spawn new processes.
-    # Right now it doesn't but it's possible that it might in the future (ex. uwsgi).
-    cmd = ["ddtrace-run", "daphne", "-p", str(SERVER_PORT), "tests.contrib.django.asgi:%s" % django_asgi]
+    env.update(django_env)
     proc = subprocess.Popen(
-        cmd,
+        django_command,
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
         close_fds=True,
         env=env,
     )
 
-    client = Client("http://localhost:%d" % SERVER_PORT)
+    client = Client("http://localhost:%s" % django_port)
 
     # Wait for the server to start up
-    client.wait()
+    try:
+        client.wait()
+    except tenacity.RetryError:
+        stdout = proc.stdout.read()
+        stderr = proc.stderr.read()
+        raise TimeoutError(
+            "Server failed to start\n======DJANGO STDOUT=====%s\n\n======DJANGO STDERR=====%s\n" % (stdout, stderr)
+        )
 
     try:
         yield client
     finally:
         resp = client.get_ignored("/shutdown-tracer")
-        assert resp.status_code == 200
+        assert resp.status_code == 200, resp.text
+        import time
+
+        time.sleep(0.5)
```

Comment:
there is a race here between the test, app and test agent.

At this point the app has submitted the traces to the test agent but there isn't a guarantee that the test agent has finished processing them by the time the test asks the test agent to perform the snapshot test.

I'll add a comment.